src/MathOptInterface.jl

```diff
@@ -77,9 +77,38 @@ Empty the instance, that is, remove the objective and all variables and constrai
 function empty! end
 
 """
-    copy!(dest::AbstractInstance, src::AbstractInstance)
+    CopyStatusCode
 
-Copy the model from the instance `src` into the instance `dest`. The target instance `dest` is emptied, and all previous indices to variables or constraints in `dest` are invalidated. Returns a dictionary-like object that translates variable and constraint indices from the `src` instance to the corresponding indices in the `dest` instance.
+An Enum of possible statuses returned by a `copy!` operation through the `CopyResult` struct.
+
+* `CopySuccess`: The copy was successful.
+* `CopyUnsupportedAttribute`: The copy failed because the destination does not support an attribute present in the source.
+* `CopyUnsupportedConstraint`: The copy failed because the destination does not support a constraint present in the source.
+* `CopyOtherError`: The copy failed for a different reason.
+
+In the failure cases, see the corresponding `message` field of the `CopyResult` for more details.
+"""
+@enum CopyStatusCode CopySuccess CopyUnsupportedAttribute CopyUnsupportedConstraint CopyOtherError
+
+"""
+    struct CopyResult{DictLike}
+        status::CopyStatusCode
+        message::String    # Human-friendly explanation why the copy failed
+        indexmap::DictLike # Only valid if status is CopySuccess
+    end
+
+A struct returned by `copy!` to indicate success or failure. If success, also exposes a map between the variable and constraint indices of the two instances.
+"""
+struct CopyResult{DictLike}
+    status::CopyStatusCode
+    message::String    # Human-friendly explanation why the copy failed
+    indexmap::DictLike # Only valid if status is CopySuccess
+end
+
+"""
+    copy!(dest::AbstractInstance, src::AbstractInstance)::CopyResult
+
+Copy the model from the instance `src` into the instance `dest`. The target instance `dest` is emptied, and all previous indices to variables or constraints in `dest` are invalidated. Returns a `CopyResult` object. If the copy is successfuly, the `CopyResult` contains a dictionary-like object that translates variable and constraint indices from the `src` instance to the corresponding indices in the `dest` instance.
```

Comment:
If the copy failed. What is the state of `dest` ? Does it need to be as it was before (without even having been emptied), empty, or anything ? I suggest, we do not impose any state, the user have just triggered a copy into `dest` so he is prepared to loose its content.

Reply:
Agreed that the state of dest should be undefined in this case. Updated the documentation.